cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -80,13 +91,114 @@ func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) fiel
 	return allErrs
 }
 
+func ValidateDiscovery(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateTLSBootstrapToken(c, fldPath)...)
+	var count int
+	if len(c.DiscoveryToken) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryToken(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryFile) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryFile(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryURL) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryURL(c, fldPath)...)
+		count++
+	}
+	if count != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
+	}
+	return allErrs
+}
+
+func ValidateTLSBootstrapToken(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.Token) != 0 && (len(cfg.TLSBootstrapToken) != 0 || len(cfg.DiscoveryToken) != 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "--token are mutually exclusive with --bootstrap-token and --discovery-token"))
+	}
+	if len(cfg.Token) == 0 && (len(cfg.TLSBootstrapToken) == 0 || len(cfg.DiscoveryToken) == 0) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(fldPath, nil, "If --token is not specified, both --bootstrap-token and --discovery-token have to be provided"),
+		)
+	}
+	if len(cfg.Token) != 0 {
+		cfg.TLSBootstrapToken = cfg.Token
+		cfg.DiscoveryToken = cfg.Token
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryFile(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !path.IsAbs(c.DiscoveryFile) {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "not an absolute file path"))
+	}
+
+	// will remove url parsing of file once Discovery struct is removed
+	u, err := url.Parse(c.DiscoveryURL)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
+	}
+
+	if len(allErrs) == 0 {
+		file.Parse(u, &c.Discovery)
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryURL(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	u, err := url.Parse(c.DiscoveryURL)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
+	}
+	if u.Scheme != "https" {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "must be https"))
+	}
+	if len(allErrs) == 0 {
+		https.Parse(u, &c.Discovery)
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryToken(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```

Comment:
Because other things still depend on the Discovery struct (which will be removed and change in a future PR) I am leaving behind all the funcs that are required to keep it working. So ` ValidateDiscoveryStruct()` and the funcs it calls are all the 'old' way, and will be updated later (will use the new functions being added).